source/common/buffer/buffer_impl.h

```diff
@@ -183,7 +186,9 @@ class Slice : public SliceData {
       copy_size = std::min(size, data_);
       data_ -= copy_size;
     }
-    memcpy(base_ + data_, src + size - copy_size, copy_size);
+    MemBlockBuilder<uint8_t> mem_builder(base_ + data_, copy_size);
```

Comment:
I think you did a nice job with this, but after looking at the result, I don't think we are really achieving that much at least in this case. The new code doesn't look any safer than the old code.

I wonder if it's better to leave it as a memcpy in this case, but use a lint exclusion as a promise that this has been well scrutizined for safety.

Reply:
Right, there are places where the use of SAFE_MEMCPY isn’t good enough because determining the size of the array cannot be achived automatically. For those cases, I end up using MemBlockBuilder, which as you said, adds more complexity to a memcpy (since in the end, the append methods uses memcpy.). Maybe, there are places where copying could be made safer by changing memcpy to memmove, but that could affect performance as well. WDYT?

Reply:
I don't think memmove helps with safety; it may just make things slower.

I was thinking you could make `BufferImpl` own a MemBlockBuilder and have the MemBlockBuilder manage its memory buffer. I think that would help, but would be more invasive, but might wind up making BufferImpl a little simpler. The idea of instantiating a MemBlockBuilder inline where you want to do a memcpy I don't think serves the goal of this exercise.

I think for surgical fixes to buffer_impl.cc I'd just use something like
```
  memcpy(base_ + data_, src + size - copy_size, copy_size); // NOLINT(safe-memcpy)
```
and make your format checker ignore that warning for lines that have that NOLINT pattern.